File: LongestSubstringWithoutRepeatingCharacters_3.java
class Solution {
    public static int lengthOfLongestSubstring(String s) {
        boolean check[] = new boolean[255] ;
        int left = 0, maxLen = 0, count = 0;
        for(int right = 0; right < s.length(); right++) {
            char c = s.charAt(right);
            if(check[c] == false) {
                check[c] = true;
                count++;
                maxLen  = Math.max(maxLen, count);
                continue;
            }
           
            if(check[c] == true) {
                while(s.charAt(left) != c) {
                    check[s.charAt(left)] = false;
                    left++;
                } 
                left++;
                count = right - left + 1;
            }
        }
        return maxLen;
    }
    
    public static void main(String[] args) {
        System.out.println(lengthOfLongestSubstring("abcabcbb"));
    }
}

File: LongestSubstringWithoutRepeatingCharacters_3.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        n = len(s)
        max_length = 0
        last_index = [0] * 128  # Mảng lưu vị trí xuất hiện cuối cùng của ký tự

        start = 0
        for end in range(n):
            current_char = s[end]
            start = max(start, last_index[ord(current_char)])
            max_length = max(max_length, end - start + 1)
            last_index[ord(current_char)] = end + 1

        return max_length

    def version2(self, s):
        char_index = {}
        start = 0
        max_length = 0

        for end, char in enumerate(s):
            if char in char_index and char_index[char] >= start:
                start = char_index[char] + 1
            char_index[char] = end
            max_length = max((max_length, end - start + 1))

        return max_length


if __name__ == "__main__":
    solution = Solution()
    print(solution.lengthOfLongestSubstring("abcabcbb"))
